internal/cloud/gcp/application_permission.go

```diff
@@ -0,0 +1,199 @@
+		if err != nil {
+			return err
+		}
+
+		modifyFunc(ctx, config.Role, config.Member, projectPolicy)
+
+		errSave := saveProjectIamPolicy(ctx, client, config.Project, projectPolicy)
+		if errSave == nil {
+			// TODO: fetch again I think?
+			// https://github.com/hashicorp/terraform-provider-google/blob/2c3be0cf1f9c56231817a2e876fa63b1afdb46e2/google/iam.go#L103
+			break
+		}
+		if isConflictError(errSave) {
+			time.Sleep(backoff)
+			backoff = backoff * 2
+			if backoff > 30*time.Second {
+				return errwrap.Wrapf(fmt.Sprintf("Error applying IAM policy to %s: Too many conflicts.  Latest error: {{err}}", "create permission"), err)
+			}
+			continue
+		}
+
+		// TODO: retry on not-found SA
+		// retry in the case that a service account is not found. This can happen when a service account is deleted
+		// out of band.
+		// if isServiceAccountNotFoundError, _ := iamServiceAccountNotFound(err); isServiceAccountNotFoundError {
+		// 	// calling a retryable function within a retry loop is not
+		// 	// strictly the _best_ idea, but this error only happens in
+		// 	// high-traffic projects anyways
+		// 	currentPolicy, rerr := iamPolicyReadWithRetry(updater)
+		// 	if rerr != nil {
+		// 		if p.Etag != currentPolicy.Etag {
+		// 			// not matching indicates that there is a new state to attempt to apply
+		// 			// log.Printf("current and old etag did not match for %s, retrying", updater.DescribeResource())
+		// 			time.Sleep(backoff)
+		// 			backoff = backoff * 2
+		// 			continue
+		// 		}
+
+		// 		// log.Printf("current and old etag matched for %s, not retrying", updater.DescribeResource())
+		// 	} else {
+		// 		// if the error is non-nil, just fall through and return the base error
+		// 		// log.Printf("[DEBUG]: error checking etag for policy %s. error: %v", updater.DescribeResource(), rerr)
+		// 	}
+		// }
+	}
+
+	return nil
+}
+
+// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/create
+func getProjectIamPolicy(ctx context.Context, service GCPResourceManagerIface, projectId string) (*cloudresourcemanager.Policy, error) {
+	tflog.Debug(ctx, "getProjectIamPolicy "+projectId)
+	getCall := service.GetIamPolicy(projectId, &cloudresourcemanager.GetIamPolicyRequest{})
+	policy, errDo := getCall.Do()
+	if errDo != nil {
+		return nil, errDo
+	}
+	tflog.Debug(ctx, "got iam policy")
+
+	return policy, nil
+}
+
+func saveProjectIamPolicy(ctx context.Context, service GCPResourceManagerIface, projectId string, policy *cloudresourcemanager.Policy) error {
+	tflog.Debug(ctx, "saveProjectIamPolicy "+projectId)
+	saveCall := service.SetIamPolicy(projectId, &cloudresourcemanager.SetIamPolicyRequest{
+		Policy: policy,
+	})
+	policy, errDo := saveCall.Do()
+	if errDo != nil {
+		return errDo
+	}
+	return nil
+}
+
+func addToPolicy(ctx context.Context, role string, member string, policy *cloudresourcemanager.Policy) error {
+	addedToExisting := false
+	for _, binding := range policy.Bindings {
```

Comment:
I'm not following the logic here. I think its because I'm not as familiar w/ GCP IAM.

A "policy" in GCP IAM is the binding of a predefined role to an entity (service account), right?

So are you iterating through all of the "policies" in the project seeing if the role already has bindings, and if so attaching another binding?

Also, I don't see you applying the condition on this - in fact I don't see the condition used anywhere in this file.

Reply:
TBH it's a little odd we have to do this. There's no API for adding a single binding and I even pawed through the google terraform provider for the iam resources and they all do this pattern of fetch and modify the project policy.

There's one policy per project and it has many bindings. The bindings have members and _a_ role. I believe that since the condition is at the role level, it would apply to all members. 

Here's a snippet of `gcloud projects get-iam-policy md-wbeebe-0808-example-app`. I'll add a condition and see if it shows up where I think it does.

```
bindings:
- members:
  - serviceAccount:gkewest-86qv-certmanager@md-wbeebe-0808-example-apps.iam.gserviceaccount.com
  role: projects/md-wbeebe-0808-example-apps/roles/gkewest86qvcertmanager
- members:
  - serviceAccount:gkewest-86qv-externaldns@md-wbeebe-0808-example-apps.iam.gserviceaccount.com
  role: projects/md-wbeebe-0808-example-apps/roles/gkewest86qvexternaldns
- members:
  - serviceAccount:example-dev-gkewest-86qv-nodes@md-wbeebe-0808-example-apps.iam.gserviceaccount.com
  role: roles/artifactregistry.reader
- members:
  - serviceAccount:service-68804004948@gcp-sa-artifactregistry.iam.gserviceaccount.com
  role: roles/artifactregistry.serviceAgent
- members:
  - serviceAccount:a-late-prefix-06@md-wbeebe-0808-example-apps.iam.gserviceaccount.com
  - serviceAccount:a-late-prefix-07@md-wbeebe-0808-example-apps.iam.gserviceaccount.com
  role: roles/cloudsql.viewer
```

Reply:
Okay so adding a condition makes a different binding 👍 

Without condition
```
- members:
  - serviceAccount:a-late-prefix-06@md-wbeebe-0808-example-apps.iam.gserviceaccount.com
  - serviceAccount:a-late-prefix-07@md-wbeebe-0808-example-apps.iam.gserviceaccount.com
  role: roles/cloudsql.viewer
```

With condition
```
- members:
  - serviceAccount:a-late-prefix-07@md-wbeebe-0808-example-apps.iam.gserviceaccount.com
  role: roles/cloudsql.viewer
- condition:
    expression: resource.type == "pubsublite.googleapis.com/Topic"
    title: 'test condition '
  members:
  - serviceAccount:a-late-prefix-06@md-wbeebe-0808-example-apps.iam.gserviceaccount.com
  role: roles/cloudsql.viewer
```